Add aten::quantize_per_tensor and aten::int_repr

mgehre-amd opened the pull request:
This PR adds support for `quantize_per_tensor` and `int_repr` and their lowering to Linalg.
The use case is to support the pattern `int_repr(quantize_per_tensor(x, scale, zero_point, type))`, which is used by the brevitas quantizer.

While working on this PR, I discovered two issues that make lowering `quantize_per_tensor` fail for `quint8` and `qint32` (`qint8` works):
(1) For uint8, I get `ERROR: dtype (torch.int8) is not equal to golden dtype (torch.uint8)` in the e2e test. It's unclear to me why this happens and how to fix. After learning more about it, I will fix it in a follow-up PR.
(2) qint32 doesn't even have a type representation in `TorchTypes.td` yet.

The strategy to lower `quantize_per_tensor` is to do `cast(clamp(round(input/scale + zero_point))` into an integer type.
Doing that, `int_repr` becomes a no-op. If we would want to implement other operators that work on quantized types like qint8,
we would also need to carry the static information (i.e. scale and zero_point) to them. For `int_repr`, this is not needed because it doesn't use that information.

*Details*:
With this PR from
```
  def forward(self, x):
   return torch.int_repr(torch.quantize_per_tensor(x, 0.5, 0, torch.qint8))
```
we first get
```
module attributes {torch.debug_module_name = "TestModule"} {
  func.func @forward(%arg0: !torch.vtensor<[3],f32>, %arg1: !torch.vtensor<[3],f32>) -> !torch.vtensor<[3],si8> {
    %float5.000000e-01 = torch.constant.float 5.000000e-01
    %int0 = torch.constant.int 0
    %int12 = torch.constant.int 12
    %0 = torch.aten.quantize_per_tensor %arg0, %float5.000000e-01, %int0, %int12 : !torch.vtensor<[3],f32>, !torch.float, !torch.int, !torch.int -> !torch.vtensor<[3],!torch.qint8>
    %1 = torch.aten.int_repr %0 : !torch.vtensor<[3],!torch.qint8> -> !torch.vtensor<[3],si8>
    return %1 : !torch.vtensor<[3],si8>
  }
}
```
and then lower that into
```
 #map = affine_map<(d0) -> (d0)>
module attributes {torch.debug_module_name = "TestModule"} {
  ml_program.global private mutable @global_seed(dense<0> : tensor<i64>) : tensor<i64>
  func.func @forward(%arg0: tensor<3xf32>, %arg1: tensor<3xf32>) -> tensor<3xi8> {
    %cst = arith.constant 0.000000e+00 : f32
    %cst_0 = arith.constant 2.550000e+02 : f32
    %0 = tensor.empty() : tensor<3xi8>
    %1 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel"]} ins(%arg0 : tensor<3xf32>) outs(%0 : tensor<3xi8>) {
    ^bb0(%in: f32, %out: i8):
      %2 = arith.addf %in, %cst : f32
      %3 = math.roundeven %2 : f32
      %4 = arith.maxf %3, %cst : f32
      %5 = arith.minf %4, %cst_0 : f32
      %6 = arith.fptoui %5 : f32 to i8
      linalg.yield %6 : i8
    } -> tensor<3xi8>
    return %1 : tensor<3xi8>
  }
}
```

ljfitz commented on python/torch_mlir_e2e_test/test_suite/quantize.py:
```diff
@@ -0,0 +1,41 @@
+    def forward(self, x):
+        y1 = torch.int_repr(torch.quantize_per_tensor(x, 1.0, 0, torch.qint8))
+        # FIXME: When lowering to TOSA, the type gets converted to int8
+        # due to convertDtypeToBuiltinElementType loosing signedness.
```
```suggestion
        # due to convertDtypeToBuiltinElementType losing signedness.
```

ljfitz commented on python/torch_mlir_e2e_test/test_suite/quantize.py:
```diff
@@ -0,0 +1,41 @@
+@register_test_case(module_factory=lambda: Quantize())
+def Quantize_basic(module, tu: TestUtils):
+    module.forward(tu.rand(3, 5, low=-10, high=10))
+    print(module.forward(tu.rand(3, 5, low=-10, high=10)))
```
Residual debugging?

mgehre-amd replied:
Oops, soory

ljfitz commented on python/torch_mlir_e2e_test/test_suite/quantize.py:
```diff
@@ -0,0 +1,41 @@
+    ])
+    def forward(self, x):
+        y1 = torch.int_repr(torch.quantize_per_tensor(x, 1.0, 0, torch.qint8))
+        # FIXME: When lowering to TOSA, the type gets converted to int8
```
This is to be fixed before upstreaming?

mgehre-amd replied:
I'm unsure. I don't want to make this PR to big, and having qint8 work is already a nice improvement.
What do you think?

ljfitz commented on lib/Conversion/TorchToTosa/TorchToTosa.cpp:
```diff
@@ -3093,6 +3093,58 @@ LogicalResult ConvertAtenOp<AtenBroadcastToOp>::matchAndRewrite(
+  auto resultTy = getTypeConverter()
+                      ->convertType(op->getOperand(0).getType());
+
+  // This is dead and will be removed
```
I'm guessing that this has to do with the pattern supported in `AtenIntReprOp`? I think cookie-trail comments are needed to help people understand and maintain the approach.

mgehre-amd replied:
Yes, I can explain this better. It feels a bit hacky, but essentially:
In `ConvertAtenOp<AtenIntReprOp>`, we replace
```
a = "torch.constant.int"() {value = 0 : i64} : () -> !torch.int
y = AtenQuantizePerTensorOp(x, offset=a)
z = AtenIntReprOp(y)
```
by 
```
a = "torch.constant.int"() {value = 0 : i64} : () -> !torch.int
y = AtenQuantizePerTensorOp(x, offset=a)
z = tosa.cast(x) // note the 'x' here
```
Now a, and y are dead. They are not removed during the TorchToTosa pass, and then it fails with
`error: failed to legalize operation 'torch.constant.int'`.

To work around this, I replace `AtenQuantizePerTensorOp` by tosa.cast, which magically gets rid of `a`,
and later this dead cast is removed by some other pass.

mgehre-amd replied:
Ok, I rewrote that part to instead:
1) replace AtenQuantizePerTensorOp by cast to int8
2) replace AtenIntReprOp by its operand (i.e. make it no-op)
This is much cleaner now.

mgehre-amd commented on lib/Conversion/TorchToTosa/TorchToTosa.cpp:
```diff
@@ -3093,6 +3093,57 @@ LogicalResult ConvertAtenOp<AtenBroadcastToOp>::matchAndRewrite(
+LogicalResult ConvertAtenOp<AtenQuantizePerTensorOp>::matchAndRewrite(
+    AtenQuantizePerTensorOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+
```
Comment for strategy of going to quantized types with TOSA

mgehre-amd replied:
Implement adds and muls

mgehre-amd commented on python/torch_mlir_e2e_test/test_suite/quantize.py:
```diff
@@ -0,0 +1,42 @@
+    ])
+    def forward(self, x):
+        y1 = torch.int_repr(torch.quantize_per_tensor(x, 1.0, 0, torch.qint8))
+        # FIXME: When lowering to TOSA, the type gets converted to int8
```
Mark XFAIL

mgehre-amd commented:
@ljfitz,
- I added a description to this PR
- removed the TOSA lowering due to the rounding issue
- implemented a linalg lowering
- in linalg, I get further with the quint8, but I'm still stuck due to `ERROR: dtype (torch.int8) is not equal to golden dtype (torch.uint8)`. I added that issue into the PR description. I would like to still go ahead and merge this to at least have the qint8 case working.

What do you think?

mgehre-amd committed:
Implement quantize_per_tensor and int_repr with their lowering to linalg

GlebKazantaev committed:
Fix OptionalCType class name (#1779)

* Fix OptionalCType class name

* Rmove LTC xfail tests

ashay committed:
CI: miscellaneous fixes for Release builds (#1781)

- Use v3 of actions/checkout, since the version we use (v2) uses
   Node.js 12, which is deprecated by GitHub.

 - Source the PowerShell venv sctipt (instead of the bash sript) since
   the calling script is a PowerShell script.  Without this, the build
   doesn't use venv at all.

 - Make the build dependencies in whl-requirements.txt (used by
   setup.py) match those in requirements.txt.  To that end, this patch
   creates a build-requirements.txt that is referenced by
   requirements.txt and whl-requirements.txt.

vivekkhandelwal1 committed:
build: manually update PyTorch version
Set PyTorch and TorchVision version to nightly release 2023-01-08.

Signed-Off By: Vivek Khandelwal<vivek@nod-labs.com>

ljfitz commented on e2e_testing/xfail_sets.py:
```diff
@@ -12,7 +12,9 @@
 
-LINALG_XFAIL_SET = COMMON_TORCH_MLIR_LOWERING_XFAILS
+LINALG_XFAIL_SET = COMMON_TORCH_MLIR_LOWERING_XFAILS.union({
+    "Quantize_uint8"
```
In your description you suggest you want help understanding why this fails. I suggest you also add that in a comment here.

ljfitz commented on lib/Conversion/TorchToLinalg/Uncategorized.cpp:
```diff
@@ -907,7 +923,8 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
 
     Value self = payloadArgs[0];
-    Value threshold = convertScalarToDtype(b, loc, adaptor.getThreshold(), dtype);
+    Value threshold =
```
This looks like an unrelated change - please see the 2nd last paragraph of https://llvm.org/docs/CodingStandards.html#introduction.

ljfitz commented on lib/Conversion/TorchToLinalg/Uncategorized.cpp:
```diff
@@ -1506,7 +1571,8 @@ class ConvertAtenDetachOp : public OpConversionPattern<AtenDetachOp> {
 
     Type resultType = getTypeConverter()->convertType(op.getType());
-    rewriter.replaceOpWithNewOp<tensor::CastOp>(op, resultType, adaptor.getSelf());
+    rewriter.replaceOpWithNewOp<tensor::CastOp>(op, resultType,
```
Unrelated reformatting?

mgehre-amd replied:
Hm, my "format changed lines" in vscode doesn't seem to always work. Sorry for missing that before review.

ljfitz commented on lib/Conversion/TorchToLinalg/Uncategorized.cpp:
```diff
@@ -929,7 +946,8 @@ static Value createLinalgPayloadCalculationForElementwiseOp(
     Value grad = convertScalarToDtype(b, loc, payloadArgs[0], dtype);
     Value self = convertScalarToDtype(b, loc, payloadArgs[1], dtype);
-    Value threshold = convertScalarToDtype(b, loc, adaptor.getThreshold(), dtype);
+    Value threshold =
```
Unrelated reformatting?

ljfitz reviewed (approved):
I'd be happy for you to use this to kick off the discussion upstream.

GlebKazantaev committed:
Added support for aten::norm.ScalarOpt_dim (#1774)

* Added support for aten::norm.ScalarOpt_dim

* Disable NormalizeModule_basic for linalg

navahgar committed:
Fix unsqueeze in Torch to Tosa conversion (#1780)

ashay committed:
build: update llvm tag to de3f0f7f (#1789)

Credit to @vivekkhandelwal1 for finding the necessary changes.

Summary of changes:

 - Switch Tosa_IntArrayAttr[N], Tosa_IntArrayAttrUpto[N] to DenseI64ArrayAttr.

 - Replace kNoIterationLimit with kNoLimit. (https://reviews.llvm.org/D140525)

 - Add dependency on MhloPasses when MHLO is enabled

 - Specify result type when using mhlo::DotOp

ashay committed:
[TOSA] Add LeakyReLU conversion pass (#1790)

* feat(TorchToTOSA): LeakyReLU legalization

* test(LeakyReLU): Add LIT test and enable e2e test

Co-authored-by: Philipp Braun <philipp.braun@amd.com>

vivekkhandelwal1 committed:
[MLIR][TORCH] Add decomposition for prims.var and prims.sqrt op

Signed-Off By: Vivek Khandelwal <vivek@nod-labs.com>

Commit:
update PyTorch version to 2.0.0.dev20230111

asaadaldien committed:
Use global for MLIRTorchOpsIncGenTdFiles (#1784)

li-plus committed:
[LINALG][TOSA][MHLO] Add e2e support for aten bitwise ops (#1753)
